nixos/modules/config/users-groups.nix

```
@@ -893,29 +893,46 @@ in
         else
           ""; # keep around for backwards compatibility
 
-      systemd.services.linger-users = lib.mkIf ((length lingeringUsers) > 0) {
+      systemd.services.linger-users = {
         wantedBy = [ "multi-user.target" ];
         after = [ "systemd-logind.service" ];
         requires = [ "systemd-logind.service" ];
 
         script =
           let
             lingerDir = "/var/lib/systemd/linger";
-            lingeringUsersFile = builtins.toFile "lingering-users" (
-              concatStrings (map (s: "${s}\n") (sort (a: b: a < b) lingeringUsers))
-            ); # this sorting is important for `comm` to work correctly
+            userPartition = lib.lists.partition (u: u.linger) (builtins.attrValues cfg.users);
+            lingeringUserNames = map (u: u.name) userPartition.right;
+            nonLingeringUserNames = map (u: u.name) userPartition.wrong;
           in
           ''
+            ${lib.strings.toShellVars { inherit lingeringUserNames nonLingeringUserNames; }}
+
+            user_configured () {
+                # Use `id` to check if the user exists rather than checking the
+                # NixOS configuration, as it may be that the user has been
+                # manually configured, which is permitted if users.mutableUsers
+                # is false (the default).
+                id "$1" >/dev/null
+            }
+
             mkdir -vp ${lingerDir}
             cd ${lingerDir}
-            for user in $(ls); do
-              if ! id "$user" >/dev/null; then
-                echo "Removing linger for missing user $user"
-                rm --force -- "$user"
-              fi
+            shopt -s dotglob nullglob
+            for user in *; do
+                if ! user_configured "$user"; then
+                    # systemd has this user configured to linger despite them not
+                    # existing.
+                    rm -- "$user"
+                fi
             done
-            ls | sort | comm -3 -1 ${lingeringUsersFile} - | xargs -r ${pkgs.systemd}/bin/loginctl disable-linger
-            ls | sort | comm -3 -2 ${lingeringUsersFile} - | xargs -r ${pkgs.systemd}/bin/loginctl  enable-linger
+
+            if (( ''${#nonLingeringUserNames[*]} > 0 )); then
+                ${pkgs.systemd}/bin/loginctl disable-linger "''${nonLingeringUserNames[@]}"
+            fi
+            if (( ''${#lingeringUserNames[*]} > 0 )); then
+                ${pkgs.systemd}/bin/loginctl enable-linger "''${lingeringUserNames[@]}"
+            fi
```

```suggestion
            if (( ''${#nonLingeringUserNames[*]} > 0 )); then
                ${lib.getExe' config.systemd.package "loginctl"} "''${nonLingeringUserNames[@]}"
            fi
            if (( ''${#lingeringUserNames[*]} > 0 )); then
                ${lib.getExe' config.systemd.package "loginctl"} enable-linger "''${lingeringUserNames[@]}"
            fi
```

Hardcoding `pkgs.systemd` here means `systemd.package` overrides would be ignored. We should use the systemd package from config instead.

---

That commit is not in here, and your current changes would very much conflict if that is in some other PR. But yes, if you want to pick that commit into here, go ahead.